### PR TITLE
Add MovingMNIST dataset

### LINKED ISSUE #6981: Add Moving MNIST dataset
### 🚀 The feature

I would like to add MovingMNIST dataset.

### Motivation, pitch

A lot of people uses Moving MNIST dataset for evaluating their proposed models according to [PaperWithCode](https://paperswithcode.com/dataset/moving-mnist).

### Alternatives

_No response_

### Additional context

I've implemented [Movning MNIST dataset](https://github.com/tsugumi-sys/SAM-ConvLSTM-Pytorch/blob/main/dataset/moving_mnist.py) in my personal project. So I would like to make a PR based on this.

I checked https://github.com/pytorch/vision/issues/3562 and https://github.com/pytorch/vision/pull/2690.

@pmeier @yassineAlouini 
From [this comment](https://github.com/pytorch/vision/pull/2690#issuecomment-1114678355) in https://github.com/pytorch/vision/pull/2690, I might should add to `torchvision/prototype/datasets/_builtin`. So where should I add Moving MNIST dataset ?

cc @pmeier

### PR DESCRIPTION
**Pull request by tsugumi-sys**
Signed-off-by: tsugumi-sys <tidemark0105@gmail.com>

Implement this class applying [this review](https://github.com/pytorch/vision/pull/2690#pullrequestreview-493281279) here.

1. Transform for each frames: Done
2. There two ways, **1:** Read form file and split. **2:** Generate dataset with this class and split.) I prefer 1 to 2 and[ tensorflow also do 1](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/video/moving_mnist.py). ([The moving sequence dataset implemented by tensorflow](https://github.com/tensorflow/datasets/blob/b9deaa07f567ad3211850a1e5c1ae993a1a03409/tensorflow_datasets/video/moving_sequence.py) is an other dataset.)
3. Simply return a flames, Not split into input and label. : Moving MNIST is used both supervised and unsupervised training,  so we should just return frames here.

closes #6981 
cc @pmeier

### CODE REVIEW

**Review by kyrianebube — APPROVED** [review]

**Comment by tsugumi-sys** [comment]
@pmeier 
Do you have any idea why these CI fails?

**Comment by pmeier** [comment]
@tsugumi-sys Both failures are unrelated.

**Comment by tsugumi-sys** [comment]
@pmeier 
I want to discuss how to split data into train and test.
We can download only a single file so we need to manually split the data. I'm considering two ways.

- Split in this class.
- This class just returns the whole data and users need to split manually.

I think **Split in this class** is not a good way because we need to load the whole data to get whether train data or test data.
Or we can save into two files when downloading the raw data, but this may be a little bit confusing. If user wants to change the split ratio, we need to download again or concatenate two files.
Do you have any idea for this?

File: test/datasets_utils.py
Comment (pmeier) [inline test/datasets_utils.py:567]: Remove before merge

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:112]: This is legacy behavior of the `MNIST` datasets. Let's store the data in `$ROOT/MovingMNIST`.

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:104]: You don't need extra error handling here

```suggestion
        download_url(
            self.url,
            root=self.raw_folder,
            filename=self.raw_filename,
            md5=self.md5,
        )
```

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:61]: After writing my initial review, we changed the video format to `(T, C, H, W)`. We should also use that here.

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:65]: Not sure what you are doing here? We **don't** want to transform the frames individually, but rather as one unit. So I guess we can simply do

```suggestion
            data = self.transform(data)
```

here?

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:25]: Since these are fixed and only used in a single location, can we inline them?

**Review by pmeier — COMMENTED** [review]
> I think **Split in this class** is not a good way because we need to load the whole data to get whether train data or test data.

Not sure I understand. You are already loading the complete file, so what is difference? Plus, we are talking about ~800MB here which is not terribly large.

My suggestion is to have two parameters: `split: Optional[str] = None` and `split_ratio: int = 10`.

1. For `split=None`, we return the full `data` without any split on our side. Meaning, we ignore `split_ratio`.
2. For `split="train"`, we return `data[:split_ratio]`.
3. For `split="test"`, we return `data[split_ratio:]`.

Here `data` is a single video in the dataset, i.e. `(T, C, H, W)`. With this the user by default gets the unsupervised data by default, but has convenient access to the splits defined by the authors.

In addition, the user can set the `split_ratio` to their liking. IMO that is enough for the first round. 

Generating new data is somewhat unrelated to `MovingMNIST` since they don't do anything special other than bouncing data in a larger image. Tensorflow has this functionality, but in a general way. If someone needs it, we can always add that later.

Thoughts?

**Comment by tsugumi-sys** [comment]
@pmeier 
Thank you for your reviews! I was warried too much. I agree your idea :)

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:51]: I think we can inline this given that it is only called once.

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:61]: Let's use the canonical terminology here and everywhere else:

```suggestion
        num_samples, num_frames, height, width = data.size()
```

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:60]: Nit: `torch.swapaxes` is an alias for `torch.transpose`

```suggestion
        data = torch.from_numpy(
            np.load(os.path.join(self.raw_folder, self.raw_filename))
        ).transpose(0, 1)
```

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [reply torchvision/datasets/moving_mnist.py:61]: Nit:

```suggestion
        batch_size, seq_len, height, width = data.shape
```

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:108]: 1. This should not be public
2. These don't need to be `@property`'s since they will always return the same thing. Thus, a regular instance attribute is sufficient
3. The term "raw" comes from  the legacy MNIST behavior and is not really applicable here. Newer datasets use the term "base folder", e.g.
   https://github.com/pytorch/vision/blob/32d254bbfcf14975f846765775584e61ef25a5bc/torchvision/datasets/country211.py#L41

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:87]: I'm sorry, I was mistaken that the URL is only used once. It could be used here to avoid repeating the file name.

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:90]: This doesn't need a docstring, since we are not going to document it in the online version and the function name is descriptive enough.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1509]: This is only used once and small enough so we can inline it.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1516]: Same as above.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1520]: This test should test the `len(dataset)` and that should always be the test equivalent of `10_000`, right? Why is there a difference between the different splits? What you are testing below should be in a separate `test_num_frames`, no?

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1501]: Let's test the extremes for `split_ratio` here as well

```suggestion
    ADDITIONAL_CONFIGS = datasets_utils.combinations_grid(split=(None, "train", "test"), split_ratio=(10, 1, 19))
```

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:41]: What about error checking for `split_ratio`?

```suggestion
        if not isinstance(split_ratio, int):
            raise TypeError(f"`split_ratio` should be an integer, but got {type(split_ratio)} instead.")
        elif not (1 <= split_ratio <= 19):
            raise ValueError(f"`split_ratio` should be `1 <= split_ratio <= 19`, but got {split_ratio} instead.")
        self.split_ratio = split_ratio
```

**Review by pmeier — COMMENTED** [review]
Thanks @tsugumi-sys! Implementation looks correct. I have a couple of inline comments to make it more compatible with the datasets that we already have.

File: torchvision/datasets/moving_mnist.py
Comment (tsugumi-sys) [reply torchvision/datasets/moving_mnist.py:41]: Thank for your reviews:)
I was thinking split ratio is used for split data into train and test dataset. but we should use it for split input and label?

File: test/test_datasets.py
Comment (tsugumi-sys) [reply test/test_datasets.py:1520]: I misunderstood of `split_ratio`.

File: torchvision/datasets/moving_mnist.py
Comment (pmeier) [inline torchvision/datasets/moving_mnist.py:18]:
```suggestion
        split_ratio (int, optional): The split ratio of number of frames. If ``split="train"``, the first split
```

**Review by pmeier — COMMENTED** [review]
Thanks for being patient @tsugumi-sys! Mostly minor nits left.

**Comment by pmeier** [comment]
Just as a heads-up for someone else reading this thread: the timestamps on the commits by @tsugumi-sys are wrong and this is why they are shown in this weird order in GH. For example https://github.com/pytorch/vision/pull/7042#pullrequestreview-1229780454 was done by me on Dec 25, 2022. eacd778db60136aa9f4499f44fb461573be3a022 was added in response to that, but is stamped with `Date: Tue, 20 Dec 2022 15:53:03 +0900`.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1504]:
```suggestion
        base_folder = os.path.join(tmpdir, self.DATASET_CLASS.__name__)
```

Comment (pmeier) [inline test/test_datasets.py:1504]:
```suggestion
        base_folder = os.path.join(tmpdir, self.DATASET_CLASS.__name__)
```

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1526]:
```suggestion
```

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1528]: Instead of checking against a zero vector, can't we simply check if all elements are zero?

```suggestion
            if config["split"] == "train":
                assert (dataset.data == 0).all()
```

Same for the test split as well.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:1541]: You had this before, but I don't think this should be methods. With my comment above, we don't need them outside of `inject_fake_data`.

**Review by pmeier — APPROVED** [review]
Thanks @tsugumi-sys! Only some minor test cleanup left, but otherwise this is good to merge if relevant CI is green. Don't worry too much about failing jobs for now, since there seem to be some unrelated failures. I'll look through them and ping you in case something needs to be fixed here.

**Comment by tsugumi-sys** [comment]
And my commit timestamp was wrong because the system datetime of my WSL2 (Ubuntu) was wrong (I dont figure out why this happened). Sorry:(

**Comment by pmeier** [comment]
None of the CI failures are related. Thanks @tsugumi-sys for the patience and good work!

**Pull request merged**
